pkg/apis/autoscaling/types.go

```diff
@@ -420,3 +420,201 @@ type HorizontalPodAutoscalerList struct {
 	// Items is the list of horizontal pod autoscaler objects.
 	Items []HorizontalPodAutoscaler
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VerticalPodAutoscalerList is a list of VerticalPodAutoscaler objects.
+type VerticalPodAutoscalerList struct {
+	metav1.TypeMeta
+	// metadata is the standard list metadata.
+	// +optional
+	metav1.ListMeta
+
+	// items is the list of vertical pod autoscaler objects.
+	Items []VerticalPodAutoscaler
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VerticalPodAutoscaler is the configuration for a vertical pod
+// autoscaler, which automatically manages pod resources based on historical and
+// real time resource utilization.
+type VerticalPodAutoscaler struct {
+	metav1.TypeMeta
+	// Standard object metadata. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Specification of the behavior of the autoscaler.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status.
+	// +optional
+	Spec VerticalPodAutoscalerSpec
+
+	// Current information about the autoscaler.
+	// +optional
+	Status VerticalPodAutoscalerStatus
+}
+
+// VerticalPodAutoscalerSpec is the specification of the behavior of the autoscaler.
+type VerticalPodAutoscalerSpec struct {
+	// A label query that determines the set of pods controlled by the Autoscaler.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	Selector *metav1.LabelSelector
+
+	// Describes the rules on how changes are applied to the pods.
+	// +optional
+	UpdatePolicy PodUpdatePolicy
+
+	// Controls how the autoscaler computes recommended resources.
+	// +optional
+	ResourcePolicy PodResourcePolicy
+}
+
+// PodUpdatePolicy describes the rules on how changes are applied to the pods.
+type PodUpdatePolicy struct {
+	// Controls when autoscaler applies changes to the pod resoures.
+	// +optional
+	UpdateMode UpdateMode
```

Comment:
Why is this optional, when the struct itself is optional above?  Put another way, is there a plausible future where the user specifies `VP.spec.updatePolicy` and provides a struct that does not include mode?

Reply:
Also, all truly optional fields should be nil-able, (pointers or slices or maps)

Reply:
PodUpdatePolicy may in future consist of multiple optional settings. It might be possible to specify PodUpdatePolicy with some other settings but leaving the mode unset (i.e. use the default).

Reg. nil-able fields: I've changed all optional fields to pointers, although I excluded "enum" string fields, because they support a natural empty value (an empty string) and I think this approach is widely used across other APIs elsewhere. WDYT?